DFE-Digital/sts-plan-technology-for-your-school
# feat: Initial redis integration

## Overview

Addresses ticket 234422; adds initial Redis integration to the web app.

## Changes

### Minor

- Adds Redis functionality

## How to review the PR

1. Add the redis connection string to the app as "ConnectionStrings:Redis"
2. Add redis caching for one of the Contentful/DB queries
3. Ensure that the query is retrieved from DB/Contentful for first request, then retrieved from Redis for subsequent requests

E.g.g to add to `GetPageQuery.cs`

```cs
public class GetPageQuery(GetPageFromContentfulQuery getPageFromContentfulQuery, GetPageFromDbQuery getPageFromDbQuery, IDistCache cache) : IGetPageQuery
{
    public async Task<Page?> GetPageBySlug(string slug, CancellationToken cancellationToken = default)
    {
        var page = await cache.GetOrCreateWithLockAsync<Page?>($"Page:{slug}", () => GetPage(slug, cancellationToken));

        return page;
    }
```

## Release requirements

None; whilst this is added to the web app, it is not currently being used anywhere, so should not cause any errors.

This can be verified by the E2E tests running successfully from this PR.

## Additional notes

1. Haven't updated documentation yet; I think we should do that once it's all in.
2. A lot of refactor/tidy up/etc. around the code Kris provided. Hope he's okay with that
3. Also accidentally _removed_ some stuff from the code Kris provided that would have been okay to have but the changes don't actually hurt. Just... code stuff. Nothing to worry about
4. There's still a lot of stuff left in that we are unlikely to use immediately. E.g. the lock provider. We _should_ use this for the Contentful webhook process to ensure only one container app runs that process at a time. We need to create a ticket for that
5. There's still a lot of stuff left in that I don't think we even need, e.g. CRUD on sets. However, code already existed so I left in.
6. 6. 😓 

## Checklist

Delete any rows that do not apply to the PR.
